lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb

```diff
@@ -0,0 +1,16 @@
+module DataUpdateScripts
+  class RemoveOrphanedAhoyRows
+    def run
+      user_ids = User.ids
+
+      # Delete all Ahoy::Events belonging to a user that does not exist anymore
+      Ahoy::Event.where.not(user_id: nil).where.not(user_id: user_ids).find_each(&:destroy)
+
+      # Delete all Ahoy::Messages belonging to a user that does not exist anymore
+      Ahoy::Message.where.not(user_id: nil).where.not(user_id: user_ids).find_each(&:destroy)
```

Comment:
There are 40255 Ahoy Messages that belong to missing users. I was able to run this in a console
```
ids = ActiveRecord::Base.connection.execute("select id from ahoy_messages where user_id is not null an
d user_id not in (select id from users);").map(&:itself);
```
However, using the user_ids with Rails notation like this in an array causes Postgres to search that whole array and even for Ahoy::Events, the query took 5+ seconds just for a count. I think to ensure this script completes we need to use straight SQL to get the job done. Then we can also ensure that our STATEMENT_TIMEOUT is bumped a little before we deploy to give the requests a little extra time to finish.

Reply:
Make sense! I'll rewrite them using SQL :-)

Reply:
Done! I wrote three SQL queries.

I wish PostgreSQL's SQL supported CTEs like this but alas it doesn't:

```sql
WITH user_ids AS (
  SELECT id from users
)

DELETE FROM ahoy_events
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)

DELETE FROM ahoy_messages
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)

DELETE FROM ahoy_visits
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)
```

Reply:
Not that it matters, but you should be able to do this with variable instead of a CTE:

```sql
DO $$
  DECLARE user_ids bigint[];
BEGIN
  user_ids := ARRAY(SELECT id from users);
  DELETE from ahoy_events
  WHERE user_id IS NOT NULL
  AND user_id != ALL(user_ids);
  
  ...
END $$;
```

Reply:
Oh nice, I didn't think of the `DO` block! TIL :D